noisy_simulator/src/instrument.rs

```diff
@@ -97,49 +98,32 @@ impl Instrument {
 }
 
 fn summed_kraus_operators(operations: &[Operation]) -> Result<Vec<SquareMatrix>, Error> {
-    // Since all the Kraus operators are square matrices of the same dimension
-    // we can cache the `vector_to_matrix_index` computation beforehand.
-    let vector_to_matrix_index = cache_vector_to_matrix_index(operations);
-
-    let choi_matrix: SquareMatrix = operations
-        .iter()
-        .map(|op| {
-            op.kraus_operators()
-                .iter()
-                .map(|k| {
-                    // This code is doing the equivalent to:
-                    // choi_matrix += vectorized(K) * vectorized(K).adjoint()
-                    // Note that if you multiply a column vector times a row
-                    // vector you get a matrix.
-                    let dim = k.shape().0.pow(2);
-                    let mut choi = SquareMatrix::zeros(dim, dim);
-                    for row in 0..dim {
-                        for col in 0..dim {
-                            choi[(row, col)] += k[vector_to_matrix_index[col]]
-                                * k[vector_to_matrix_index[row]].conj();
-                        }
-                    }
-                    choi
-                })
-                .sum::<SquareMatrix>()
-        })
-        .sum();
-
+    let choi_matrix = compute_choi_matrix(operations);
     let (choi_dim, _) = choi_matrix.shape();
     let eigen_decomposition = choi_matrix.symmetric_eigen();
     let eigenvectors = eigen_decomposition.eigenvectors;
     let eigenvalues = eigen_decomposition.eigenvalues;
     let mut summed_kraus_operators = Vec::new();
```

Comment:
Given that an eigenvalue <= 0 is an error case, would it make sense to preallocate the summed_kraus_operateors vector to the length of the eigenvalues?

Reply:
Yes, you are right.